neuland/digitizing/DigitizingEngine.h

```diff
@@ -35,46 +37,127 @@ namespace Neuland
+            mutable Validated<std::vector<Signal>> fSignals; // output signals from the channel
+            Paddle* fPaddle; // pointer to the paddle who owns this channel
+
+          private:
+            virtual void ConstructSignals() const = 0;
         };
 
         class Paddle
         {
           public:
-            Paddle(std::unique_ptr<Channel> l, std::unique_ptr<Channel> r);
+            template <class T>
+            struct Pair
+            {
+                T left{};
+                T right{};
+                Pair() = default;
+                Pair(T l, T r)
+                    : left(l)
+                    , right(r){};
+            };
+
+            struct Signal
+            {
+                Double_t energy{};
+                Double_t time{};
+                Double_t position{};
+                Pair<Double_t> channelE{};
+                Pair<Double_t> channelTdc{};
+                Pair<Double_t> channelQdc{};
+                Signal() = default;
+                Signal(Double_t e, Double_t t)
+                    : energy(e)
+                    , time(t)
+                {
+                }
+            };
+
+            Paddle(const Int_t paddleID,
+                   std::unique_ptr<Channel> l,
+                   std::unique_ptr<Channel> r,
+                   R3BNeulandHitPar* par = nullptr);
+            virtual ~Paddle() {}
             void DepositLight(Double_t time, Double_t light, Double_t dist);
 
             bool HasFired() const;
             bool HasHalfFired() const;
-            Double_t GetEnergy(UShort_t index = 0) const;
-            Double_t GetTime(UShort_t index = 0) const;
-            Double_t GetPosition(UShort_t index = 0) const;
-            const UShort_t GetNHits() const;
+
+            void SetHitModulePar(R3BNeulandHitPar* par);
+            R3BNeulandHitModulePar* GetHitModulePar() const { return fHitModulePar; }
+
+            Int_t GetPaddleId() const { return fPaddleId; }
+            const std::vector<Signal>& GetSignals() const;
 
             const Channel* GetLeftChannel() const { return fLeftChannel.get(); }
             const Channel* GetRightChannel() const { return fRightChannel.get(); }
 
           protected:
+            using ChannelSignals = std::vector<Channel::Signal>;
+            using PaddleSignals = std::vector<Signal>;
             std::unique_ptr<Channel> fLeftChannel;
             std::unique_ptr<Channel> fRightChannel;
+            R3BNeulandHitModulePar* fHitModulePar = nullptr;
+            mutable Validated<PaddleSignals> fSignals;
+            const Int_t fPaddleId;
 
           public:
             static constexpr Double_t gHalfLength = 135.;   // [cm]
             static constexpr Double_t gCMedium = 14.;       // speed of light in material in [cm/ns]
             static constexpr Double_t gAttenuation = 0.008; // light attenuation of plastic scintillator [1/cm]
             static constexpr Double_t gLambda = 1. / 2.1;
+
+          private:
```

Comment:
Because anyone who inherits this class, if necessary, could define his own CompareSignals or ConstructPaddleSignals using the data from two channels. If not, he can just leave it and doesn't need those functions.

Another reason I read somewhere is the virtual method of a class should better be private:
https://stackoverflow.com/questions/2170688/private-virtual-method-in-c